packages/core/src/amazonqFeatureDev/controllers/chat/controller.ts

```diff
@@ -439,25 +452,82 @@ export class FeatureDevController {
-            this.messenger.sendAsyncEventProgress(tabID, false, undefined)
-
-            // Lock the chat input until they explicitly click one of the follow ups
-            this.messenger.sendChatInputEnabled(tabID, false)
 
-            if (!this.isAmazonQVisible) {
-                const open = 'Open chat'
-                const resp = await vscode.window.showInformationMessage(
-                    i18n('AWS.amazonq.featureDev.answer.qGeneratedCode'),
-                    open
+            if (session?.state?.tokenSource?.token.isCancellationRequested) {
+                this.workOnNewTask(
+                    session,
+                    session.state.codeGenerationRemainingIterationCount ||
+                        TotalSteps - (session.state?.currentIteration || 0),
+                    session.state.codeGenerationTotalIterationCount || TotalSteps,
+                    session?.state?.tokenSource?.token.isCancellationRequested
                 )
-                if (resp === open) {
-                    await vscode.commands.executeCommand('aws.AmazonQChatView.focus')
-                    // TODO add focusing on the specific tab once that's implemented
+                this.disposeToken(session)
+            } else {
+                this.messenger.sendAsyncEventProgress(tabID, false, undefined)
+
+                // Lock the chat input until they explicitly click one of the follow ups
+                this.messenger.sendChatInputEnabled(tabID, false)
+
+                if (!this.isAmazonQVisible) {
+                    const open = 'Open chat'
+                    const resp = await vscode.window.showInformationMessage(
+                        i18n('AWS.amazonq.featureDev.answer.qGeneratedCode'),
+                        open
+                    )
+                    if (resp === open) {
+                        await vscode.commands.executeCommand('aws.AmazonQChatView.focus')
+                        // TODO add focusing on the specific tab once that's implemented
+                    }
                 }
             }
         }
     }
+    private workOnNewTask(
+        message: any,
+        remainingIterations?: number,
+        totalIterations?: number,
+        isStoppedGeneration: boolean = false
+    ) {
+        if (isStoppedGeneration) {
+            this.messenger.sendAnswer({
+                message:
+                    (remainingIterations ?? 0) <= 0
+                        ? "I stopped generating your code. You don't have more iterations left, however, you can start a new session."
+                        : `I stopped generating your code. If you want to continue working on this task, provide another description. You have ${remainingIterations} out of ${totalIterations} code generations left.`,
+                type: 'answer-part',
+                tabID: message.tabID,
+            })
+        }
+
+        if (((remainingIterations ?? 0) <= 0 && isStoppedGeneration) || !isStoppedGeneration) {
```

Comment:
If remainingIterations is 0 when undefined can we just add that into the function definition `remainingIterations: number = 0`

Reply:
Good call, I'll update